File: tools/market_map/path_probability.js
/* path_probability.js — JS port of path_probability.py (path-dependent cone probabilities).
 *
 * Vanilla ES5-ish, no deps. Locked to the Python reference via the committed golden fixture
 * path_probability_golden.json (Python is authoritative); test_path_probability_parity.mjs checks
 * JS == Py to 1e-6. All quantities are in LOG-RETURN space: barrier b = ln(B/S0), scale s = sigmaDaily
 * * sqrt(T), drift m = muDaily * T. Lets the terminal show the exact closed-form touch odds, MFE/MAE and
 * P(end>=level | touched barrier) live, and cross-check its Monte-Carlo cone.
 */
(function (root, factory) {
  var mod = factory();
  if (typeof module === 'object' && module.exports) module.exports = mod;
  else root.PathProb = mod;
})(typeof self !== 'undefined' ? self : this, function () {
  var SQRT2 = Math.sqrt(2.0), SQRT_2_OVER_PI = Math.sqrt(2.0 / Math.PI);

  function erf(x) { // Abramowitz-Stegun 7.1.26 (|err|<1.5e-7)
    var s = x < 0 ? -1 : 1; x = Math.abs(x);
    var t = 1.0 / (1.0 + 0.3275911 * x);
    var y = 1.0 - (((((1.061405429 * t - 1.453152027) * t) + 1.421413741) * t - 0.284496736) * t + 0.254829592) * t * Math.exp(-x * x);
    return s * y;
  }
  function ncdf(x) { return 0.5 * (1.0 + erf(x / SQRT2)); }
  function nppf(p) { // Acklam inverse-normal
    if (p <= 0) return -Infinity; if (p >= 1) return Infinity;
    var a = [-3.969683028665376e+01, 2.209460984245205e+02, -2.759285104469687e+02, 1.383577518672690e+02, -3.066479806614716e+01, 2.506628277459239e+00];
    var b = [-5.447609879822406e+01, 1.615858368580409e+02, -1.556989798598866e+02, 6.680131188771972e+01, -1.328068155288572e+01];
    var c = [-7.784894002430293e-03, -3.223964580411365e-01, -2.400758277161838e+00, -2.549732539343734e+00, 4.374664141464968e+00, 2.938163982698783e+00];
    var d = [7.784695709041462e-03, 3.224671290700398e-01, 2.445134137142996e+00, 3.754408661907416e+00];
    var plow = 0.02425, phigh = 1 - plow, q, r;
    if (p < plow) { q = Math.sqrt(-2 * Math.log(p)); return (((((c[0] * q + c[1]) * q + c[2]) * q + c[3]) * q + c[4]) * q + c[5]) / ((((d[0] * q + d[1]) * q + d[2]) * q + d[3]) * q + 1); }
    if (p > phigh) { q = Math.sqrt(-2 * Math.log(1 - p)); return -(((((c[0] * q + c[1]) * q + c[2]) * q + c[3]) * q + c[4]) * q + c[5]) / ((((d[0] * q + d[1]) * q + d[2]) * q + d[3]) * q + 1); }
    q = p - 0.5; r = q * q; return (((((a[0] * r + a[1]) * r + a[2]) * r + a[3]) * r + a[4]) * r + a[5]) * q / (((((b[0] * r + b[1]) * r + b[2]) * r + b[3]) * r + b[4]) * r + 1);
  }

  function priceToReturn(price, s0) { return Math.log(price / s0); }
  function returnToPrice(r, s0) { return s0 * Math.exp(r); }

  function touchUp(b, s, m) {
    m = m || 0.0;
    if (s <= 0) return b <= 0 ? 1.0 : 0.0;
    if (b <= 0) return 1.0;
    return Math.min(1.0, ncdf((m - b) / s) + Math.exp(2.0 * m * b / (s * s)) * ncdf((-b - m) / s));
  }
  function touchDown(b, s, m) { return touchUp(-b, s, -(m || 0.0)); }
  function runningMaxCdfGe(b, s, m) { return b > 0 ? touchUp(b, s, m) : 1.0; }

  function runningMaxQuantile(p, s, m) {
    m = m || 0.0;
    if (s <= 0) return Math.max(0.0, m);
    if (Math.abs(m) < 1e-15) return s * nppf((1.0 + p) / 2.0);
    var lo = 0.0, hi = m + 12.0 * s, mid, i;
    for (i = 0; i < 80; i++) { mid = 0.5 * (lo + hi); if ((1.0 - runningMaxCdfGe(mid, s, m)) < p) lo = mid; else hi = mid; }
    return 0.5 * (lo + hi);
  }

  function expectedMaxFavorable(s, m) {
    m = m || 0.0;
    if (s <= 0) return Math.max(0.0, m);
    if (Math.abs(m) < 1e-15) return s * SQRT_2_OVER_PI;
    var hi = Math.max(0.0, m) + 12.0 * s, n = 2000, h = hi / n, i;
    var tot = runningMaxCdfGe(0.0, s, m) + runningMaxCdfGe(hi, s, m);
    for (i = 1; i < n; i++) tot += (i % 2 ? 4 : 2) * runningMaxCdfGe(i * h, s, m);
    return tot * h / 3.0;
  }
  function expectedMaxAdverse(s, m) { return expectedMaxFavorable(s, -(m || 0.0)); }

  function probEndAboveGivenTouchUp(b, k, s, m) {
    m = m || 0.0;
    if (s <= 0) return (m >= b && m >= k) ? 1.0 : 0.0;
    var denom = touchUp(b, s, m);
    if (denom <= 0) return 0.0;
    // exact reflection for m==0 (the terminal path_report uses driftless)
    var joint = (k <= b) ? (2.0 * ncdf(-b / s) - ncdf(-(2.0 * b - k) / s)) : ncdf(-k / s);
    return Math.max(0.0, Math.min(1.0, joint / denom));
  }

  function _r(x, n) { var p = Math.pow(10, n); return Math.round(x * p) / p; }  // match Python round()
  function pathReport(s0, sigmaDaily, T, barrierUp, barrierDn, level, driftDaily) {
    var s = sigmaDaily * Math.sqrt(T), m = (driftDaily || 0.0) * T;
    // rounding mirrors path_probability.py's path_report exactly (ret 6dp, price 4dp, prob 4dp) so the
    // committed golden fixture matches to floating precision.
    var out = { T: T, s: _r(s, 6), m: _r(m, 6), mfeRet: _r(expectedMaxFavorable(s, m), 6), maeRet: _r(expectedMaxAdverse(s, m), 6) };
    out.mfePrice = _r(returnToPrice(out.mfeRet, s0), 4);
    out.maePrice = _r(returnToPrice(-out.maeRet, s0), 4);
    if (barrierUp != null && barrierUp > s0) out.touchUp = _r(touchUp(priceToReturn(barrierUp, s0), s, m), 4);
    if (barrierDn != null && barrierDn > 0 && barrierDn < s0) out.touchDn = _r(touchDown(priceToReturn(barrierDn, s0), s, m), 4);
    if (barrierUp != null && level != null && barrierUp > s0)
      out.pEndAboveGivenTouchUp = _r(probEndAboveGivenTouchUp(priceToReturn(barrierUp, s0), priceToReturn(level, s0), s, m), 4);
    return out;
  }

  return {
    ncdf: ncdf, nppf: nppf, priceToReturn: priceToReturn, returnToPrice: returnToPrice,
    touchUp: touchUp, touchDown: touchDown, runningMaxQuantile: runningMaxQuantile,
    expectedMaxFavorable: expectedMaxFavorable, expectedMaxAdverse: expectedMaxAdverse,
    probEndAboveGivenTouchUp: probEndAboveGivenTouchUp, pathReport: pathReport
  };
});

File: tools/market_map/path_probability.py
#!/usr/bin/env python3
"""path_probability.py — path-dependent forecast probabilities for the cone (pure stdlib).

The cone reports flag path-dependence as the biggest missing piece: a terminal cone "is blind to how it
gets there." lineage.py already has the Brownian-bridge barrier correction and reflection first-passage;
this module adds the rest of the path suite the Institutional Blueprint lists — expected MAXIMUM FAVORABLE
and MAXIMUM ADVERSE EXCURSION, the running-max/min distribution and quantiles, and P(end above a level |
first touched a barrier) — with exact closed forms for the driftless case and a drift-capable Monte-Carlo
that doubles as the verification oracle.

All quantities are in LOG-RETURN space relative to S0 (barrier b = ln(B/S0), level k = ln(K/S0)); the
total-horizon scale is s = sigma_daily * sqrt(T) and drift m = mu_daily * T. Convert to price with
price_to_return()/return_to_price(). Every closed form is checked against Monte-Carlo in test_path_probability.py.
"""
from __future__ import annotations

import math
import random

SQRT2 = math.sqrt(2.0)
SQRT_2_OVER_PI = math.sqrt(2.0 / math.pi)


def _ncdf(x):
    return 0.5 * math.erfc(-x / SQRT2)


def _npdf(x):
    return math.exp(-0.5 * x * x) / math.sqrt(2.0 * math.pi)


def _nppf(p):
    if p <= 0:
        return -math.inf
    if p >= 1:
        return math.inf
    a = [-3.969683028665376e+01, 2.209460984245205e+02, -2.759285104469687e+02,
         1.383577518672690e+02, -3.066479806614716e+01, 2.506628277459239e+00]
    b = [-5.447609879822406e+01, 1.615858368580409e+02, -1.556989798598866e+02,
         6.680131188771972e+01, -1.328068155288572e+01]
    c = [-7.784894002430293e-03, -3.223964580411365e-01, -2.400758277161838e+00,
         -2.549732539343734e+00, 4.374664141464968e+00, 2.938163982698783e+00]
    d = [7.784695709041462e-03, 3.224671290700398e-01, 2.445134137142996e+00, 3.754408661907416e+00]
    plow, phigh = 0.02425, 1 - 0.02425
    if p < plow:
        q = math.sqrt(-2 * math.log(p))
        return (((((c[0]*q+c[1])*q+c[2])*q+c[3])*q+c[4])*q+c[5]) / ((((d[0]*q+d[1])*q+d[2])*q+d[3])*q+1)
    if p > phigh:
        q = math.sqrt(-2 * math.log(1 - p))
        return -(((((c[0]*q+c[1])*q+c[2])*q+c[3])*q+c[4])*q+c[5]) / ((((d[0]*q+d[1])*q+d[2])*q+d[3])*q+1)
    q = p - 0.5
    r = q * q
    return (((((a[0]*r+a[1])*r+a[2])*r+a[3])*r+a[4])*r+a[5])*q / (((((b[0]*r+b[1])*r+b[2])*r+b[3])*r+b[4])*r+1)


def price_to_return(price, s0):
    return math.log(price / s0)


def return_to_price(r, s0):
    return s0 * math.exp(r)


# ---- first-passage (running max/min crossing a barrier) --------------------------------------
def touch_up(b, s, m=0.0):
    """P( max_{t<=T} X_t >= b ) for arithmetic BM X ~ (drift m, scale s), b > 0. Driftless -> 2*Phi(-b/s)."""
    if s <= 0:
        return 1.0 if b <= 0 else 0.0
    if b <= 0:
        return 1.0
    return min(1.0, _ncdf((m - b) / s) + math.exp(2.0 * m * b / (s * s)) * _ncdf((-b - m) / s))


def touch_down(b, s, m=0.0):
    """P( min_{t<=T} X_t <= b ) for b < 0 (barrier below start). By symmetry: touch_up(-b, s, -m)."""
    return touch_up(-b, s, -m)


# ---- running-max / -min distribution + quantiles + expected excursions -----------------------
def running_max_cdf_ge(b, s, m=0.0):
    """P(M >= b) where M = running max (b can be any real; <=0 -> 1)."""
    return touch_up(b, s, m) if b > 0 else 1.0


def running_max_quantile(p, s, m=0.0):
    """p-quantile of the running max M. Driftless closed form: s * Phi^{-1}((1+p)/2). Drift: invert the CDF."""
    if s <= 0:
        return max(0.0, m)
    if abs(m) < 1e-15:
        return s * _nppf((1.0 + p) / 2.0)
    lo, hi = 0.0, m + 12.0 * s
    for _ in range(80):
        mid = 0.5 * (lo + hi)
        if (1.0 - running_max_cdf_ge(mid, s, m)) < p:   # P(M < mid)
            lo = mid
        else:
            hi = mid
    return 0.5 * (lo + hi)


def expected_max_favorable(s, m=0.0):
    """E[ running MAX ] over [0,T] — Maximum Favorable Excursion. Driftless: s*sqrt(2/pi)."""
    if s <= 0:
        return max(0.0, m)
    if abs(m) < 1e-15:
        return s * SQRT_2_OVER_PI
    # E[M] = integral_0^inf P(M>=b) db  (numerical Simpson; integrand decays like a Gaussian tail)
    hi = max(0.0, m) + 12.0 * s
    n = 2000
    h = hi / n
    tot = running_max_cdf_ge(0.0, s, m) + running_max_cdf_ge(hi, s, m)
    for i in range(1, n):
        tot += (4 if i % 2 else 2) * running_max_cdf_ge(i * h, s, m)
    return tot * h / 3.0


def expected_max_adverse(s, m=0.0):
    """E[ |running MIN| ] over [0,T] — Maximum Adverse Excursion (positive magnitude). By symmetry of the
    running min: min X (drift m) =_d -max(-X) (drift -m), so MAE = expected_max_favorable(s, -m)."""
    return expected_max_favorable(s, -m)


# ---- conditional: P(end above level K | first touched barrier B) -----------------------------
def prob_end_above_given_touch_up(b, k, s, m=0.0, mc=None):
    """P( X_T >= k | max_{t<=T} X_t >= b ), b > 0. Exact reflection for m==0; MC for drift (mc = n paths)."""
    if s <= 0:
        return 1.0 if (m >= b and m >= k) else 0.0
    denom = touch_up(b, s, m)
    if denom <= 0:
        return 0.0
    if abs(m) < 1e-15:
        # joint P(M>=b, X_T>=k):  k<=b -> 2*Phi(-b/s) - Phi(-(2b-k)/s);  k>b -> Phi(-k/s)
        joint = (2.0 * _ncdf(-b / s) - _ncdf(-(2.0 * b - k) / s)) if k <= b else _ncdf(-k / s)
        return max(0.0, min(1.0, joint / denom))
    n = int(mc or 40000)
    hit = end_above_and_hit = 0
    for _ in range(n):
        x, mx = 0.0, 0.0
        steps = 64
        dt = 1.0 / steps
        sd = s * math.sqrt(dt)
        md = m * dt
        for _ in range(steps):
            x += md + sd * random.gauss(0, 1)
            if x > mx:
                mx = x
        if mx >= b:
            hit += 1
            if x >= k:
                end_above_and_hit += 1
    return (end_above_and_hit / hit) if hit else 0.0


def path_report(s0, sigma_daily, T, barrier_up=None, barrier_dn=None, level=None, drift_daily=0.0) -> dict:
    """Assemble the path-dependent panel for one horizon in PRICE space. Returns MFE/MAE (as price moves),
    touch odds to the given barriers, and P(end above level | touched the up barrier)."""
    s = sigma_daily * math.sqrt(T)
    m = drift_daily * T
    out = {"T": T, "s": round(s, 6), "m": round(m, 6),
           "mfeRet": round(expected_max_favorable(s, m), 6),
           "maeRet": round(expected_max_adverse(s, m), 6)}
    out["mfePrice"] = round(return_to_price(out["mfeRet"], s0), 4)
    out["maePrice"] = round(return_to_price(-out["maeRet"], s0), 4)
    if barrier_up is not None and barrier_up > s0:
        out["touchUp"] = round(touch_up(price_to_return(barrier_up, s0), s, m), 4)
    if barrier_dn is not None and 0 < barrier_dn < s0:
        out["touchDn"] = round(touch_down(price_to_return(barrier_dn, s0), s, m), 4)
    if barrier_up is not None and level is not None and barrier_up > s0:
        out["pEndAboveGivenTouchUp"] = round(
            prob_end_above_given_touch_up(price_to_return(barrier_up, s0), price_to_return(level, s0), s, m), 4)
    return out
